### PR TITLE
SONARJAVA-2061 Stack and consume correctly exceptional symbolic values

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 1]
```diff
@@ -725,7 +729,19 @@ private void executeVariable(VariableTree variableTree, @Nullable Tree terminato
...
         sv = SymbolicValue.FALSE_LITERAL;
       } else if (variableTree.parent().is(Tree.Kind.CATCH)) {
-        sv = constraintManager.createSymbolicValue(variableTree);
+        Type caughtType = variableTree.symbol().type();
+        Type exceptionType = null;
+        if (programState.peekValue() instanceof SymbolicValue.ExceptionalSymbolicValue) {
+          ProgramState.Pop unstack = programState.unstackValue(1);
+          programState = unstack.state;
+          // use the Exceptional SV from the stack
+          sv = unstack.values.get(0);
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [CFG.java#L922](https://github.com/SonarSource/sonar-java/blob/36a16c88170f8ec65ad7308cb0d9fd4f69fbc098/java-frontend/src/main/java/org/sonar/java/cfg/CFG.java#L922): Make this line start at column 7. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AIndentationCheck)

[CRITICAL]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 2]
```diff
@@ -725,7 +729,19 @@ private void executeVariable(VariableTree variableTree, @Nullable Tree terminato
       } else if (variableTree.type().symbolType().is("boolean")) {
         sv = SymbolicValue.FALSE_LITERAL;
       } else if (variableTree.parent().is(Tree.Kind.CATCH)) {
-        sv = constraintManager.createSymbolicValue(variableTree);
+        Type caughtType = variableTree.symbol().type();
+        Type exceptionType = null;
+        if (programState.peekValue() instanceof SymbolicValue.ExceptionalSymbolicValue) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: This if is actually not really required and more a revelation of some issue where we lack a correct stacking of SV. Please at the very least add a comment regarding this and create a ticket to handle cases where this condition is not satisfied.

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 3]
```diff
@@ -755,6 +760,24 @@ private void executeVariable(VariableTree variableTree, @Nullable Tree terminato
     }
   }
 
+  private SymbolicValue getCaughtException(Type caughtType) {
+    SymbolicValue sv = null;
+    Type exceptionType = null;
+    // FIXME SONARJAVA-2069 every path conducting to a catch block should have an exceptional symbolic value on top of the stack
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [CFG.java#L922](https://github.com/SonarSource/sonar-java/blob/da5c2c32c56c916baa904bc87a0232b4fbbba830/java-frontend/src/main/java/org/sonar/java/cfg/CFG.java#L922): Make this line start at column 7. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AIndentationCheck)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Pull request merged**